ibis/common/patterns.py

```diff
@@ -275,6 +294,160 @@ def __rich_repr__(self):
             yield name, getattr(self, name)
 
 
+class Creator(Hashable):
```

Comment:
This needs a name that is more descriptive, or at least a docstring describing what concept it's modeling.

Reply:
Yeah, I'm not satisfied with it either. Going to add docstring, in the meantime I'm open for suggestions. 

There is a single place where creator objects can be used, the right hand side of the replace pattern: `Replace(pattern, creator)`. Semantically when a match occurs for the pattern, the context contains all the captured values and information relevant for the creator, and the creator construct a new object only given by the context.

Reply:
I was think `Replacer`, `Applier`, `Substitutor`, but the replacement is actually done by the `Replace` pattern and `Creator` objects just construct things given the context.

Reply:
I renamed it to Builder, still not satisfied but we can change it later.